File: ios/RNRDocScannerViewManager.m
#import <React/RCTBridge.h>
#import <React/RCTUIManager.h>
#import <React/RCTViewManager.h>

#import "react-native-rectangle-doc-scanner-Swift.h"

@interface RCT_EXTERN_MODULE(RNRDocScannerViewManager, RCTViewManager)
RCT_EXPORT_VIEW_PROPERTY(detectionCountBeforeCapture, NSNumber)
RCT_EXPORT_VIEW_PROPERTY(autoCapture, BOOL)
RCT_EXPORT_VIEW_PROPERTY(enableTorch, BOOL)
RCT_EXPORT_VIEW_PROPERTY(quality, NSNumber)
RCT_EXPORT_VIEW_PROPERTY(useBase64, BOOL)
RCT_EXPORT_VIEW_PROPERTY(onRectangleDetect, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onPictureTaken, RCTDirectEventBlock)

RCT_EXTERN_METHOD(capture:(nonnull NSNumber *)reactTag
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(reset:(nonnull NSNumber *)reactTag)
@end

File: ios/RNRDocScannerViewManager.swift
import Foundation
import React

@objc(RNRDocScannerViewManager)
class RNRDocScannerViewManager: RCTViewManager {
  override static func requiresMainQueueSetup() -> Bool {
    true
  }

  override func view() -> UIView! {
    RNRDocScannerView()
  }

  @objc func capture(_ reactTag: NSNumber, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    bridge.uiManager.addUIBlock { _, viewRegistry in
      guard let view = viewRegistry?[reactTag] as? RNRDocScannerView else {
        reject(RNRDocScannerError.viewNotFound.code, RNRDocScannerError.viewNotFound.message, nil)
        return
      }

      view.capture { result in
        switch result {
        case let .success(payload):
          resolve([
            "croppedImage": payload.croppedImage ?? NSNull(),
            "initialImage": payload.originalImage,
            "width": payload.width,
            "height": payload.height,
          ])
        case let .failure(error as RNRDocScannerError):
          reject(error.code, error.message, error)
        case let .failure(error):
          reject("capture_failed", error.localizedDescription, error)
        }
      }
    }
  }

  @objc func reset(_ reactTag: NSNumber) {
    bridge.uiManager.addUIBlock { _, viewRegistry in
      guard let view = viewRegistry?[reactTag] as? RNRDocScannerView else {
        return
      }
      view.resetStability()
    }
  }
}
